railgun/manager/run.go

```diff
@@ -286,6 +310,50 @@ func Run(ctx context.Context, c *config.Config) error {
 	} else {
 		setupLog.Info("anonymous reports disabled, skipping")
 	}
+
+	go FlipKnativeController(mgr, prx, &c.KnativeIngressEnabled, c, setupLog)
 	setupLog.Info("starting manager")
 	return mgr.Start(ctx)
 }
+
+// wait for knative cr before register and starting knative controller
+func FlipKnativeController(mgr manager.Manager, prx proxy.Proxy, enablestatus *util.EnablementStatus, cfg *config.Config, log logr.Logger) error {
```

Comment:
Fundamentally I like what we're trying to do here: adding watchers which can proactively identify APIs that are added and hotload the controllers is a good feature.

However, I think we will need to do more to support this feature, the implementation so far has a few troubles and it also seems like something we're going to have to standardize further on. There's some growth of scope for KIC 2.0 which I don't feel we should be taking on just yet.

My recommendation:

1) add logging to the controller manager to inform the operator of decisions made to automatically disable the Knative Ingress controller
2) drop this function
3) create a follow up issue for us to discuss and groom which suggests this functionality for a separate scope

Reply:
This mechanism dynamically detect newly deployed CRD/CR after KIC is UP, and purposedly notify setup manager using registration and event setup.

 It is dedicated go routines that bump up controllers dynamically. we we can use the mechanism  dropping current pre-requisite CRD deployments such as KongClusterPlugin, KongConsumer, KongIngress, KongPlugin, TCPIngress, UDPIngress if there is restrict requirement from the cluster KIC runs on.

follow up issues for better log/convention/performance tests could be created for discussion,  they are awesome ideas.

Reply:
I don't feel like this was actually ready to be resolved as I still have some lingering concerns with the current implementation, however as this PR has been stuck in a "limbo" for some time I have created a follow up where we can look further into this so that it does not block current progress:

https://github.com/Kong/kubernetes-ingress-controller/issues/1449